tools/validate_pip_freeze.py

```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env python3
+# Verify that the installed Python modules match the requirements file provided
+# as the argument.
+import logging
+import os.path
+import pkg_resources
+import re
+import subprocess
+import sys
+import urllib.parse
+
+
+_IGNORED_WHITELIST = [
+    'python-apt',  # installed by apt from a deb
+]
+_HINT_RE = re.compile(r'.*validator-hint:( +name=(?P<name>[\w.-]+))?( +version=(?P<version>[\w.+-]+))? *$')
+
+
+def _duplicates(strings: [str]) -> [str]:
+    counts = {}
+    for s in strings:
+        if s not in counts:
+            counts[s] = 0
+        counts[s] += 1
+    return [s for s, v in counts.items() if v > 1]
+
+
+def _extract_info_from_url(line):
+    parsed = urllib.parse.urlparse(line.strip())
+    base = os.path.basename(parsed.path)
+    name = os.path.splitext(base)[0]
+
+    m = _HINT_RE.match(parsed.fragment)
+    if m:
+        if m.group('name'):
+            name = m.group('name')
+        version = m.group('version')
+        if version:
+            if version == 'SNAPSHOT':
+                return '%s===%s' % (name, version)
+            else:
+                return '%s==%s' % (name, version)
+    else:
```

Comment:
Is the `else:` here correct? We will return `None` if not `version` is specified in the hint.

Reply:
Good catch, fixed!